src/libraries/Microsoft.PowerFx.Core/Public/Values/FormulaValueNewJson.cs

```diff
@@ -0,0 +1,140 @@
+                using var jsonMemStream = new MemoryStream();
+                using var paJsonWriter = new Utf8JsonWriter(jsonMemStream);
+                var propBag = document.RootElement;
+
+                return FromJson(propBag);
+            }
+            catch
+            {
+                // $$$ better error handling here?
+                throw;
+            }
+        }
+
+        public static FormulaValue FromJson(JsonElement element)
+        {
+            switch (element.ValueKind)
+            {
+                case JsonValueKind.Null:
+                    return new BlankValue(IRContext.NotInSource(FormulaType.Blank));
+
+                case JsonValueKind.Number:
+                    return new NumberValue(IRContext.NotInSource(FormulaType.Number), element.GetDouble());
+
+                case JsonValueKind.String:
+                    return new StringValue(IRContext.NotInSource(FormulaType.String), element.GetString());
+
+                case JsonValueKind.False:
+                    return new BooleanValue(IRContext.NotInSource(FormulaType.Boolean), false);
+
+                case JsonValueKind.True:
+                    return new BooleanValue(IRContext.NotInSource(FormulaType.Boolean), true);
+
+                case JsonValueKind.Object:
+                    return RecordFromJsonObject(element);
+
+                case JsonValueKind.Array:
+                    return TableFromJsonArray(element);
+
+                default:
+                    throw new NotImplementedException($"Unrecognized JsonElement {element.ValueKind}");
+            }
+        }
+
+        // Json objects parse to records. 
+        internal static RecordValue RecordFromJsonObject(JsonElement element)
+        {
+            Contract.Assert(element.ValueKind == JsonValueKind.Object);
+
+            var fields = new List<NamedValue>();
+            var type = new RecordType();
+
+            foreach (var pair in element.EnumerateObject())
+            {
+                var name = pair.Name;
+                var value = pair.Value;
+
+                var paValue = FromJson(value);
+                fields.Add(new NamedValue(name, paValue));
+                type = type.Add(new NamedFormulaType(name, paValue.IRContext.ResultType));
+            }
+
+            return new InMemoryRecordValue(IRContext.NotInSource(type), fields);
+        }
+
+        // More type safe than base class's ParseJson
+        // Parse json. 
+        // [1,2,3]  is a single column table, actually equivalent to: 
+        // [{Value : 1, Value: 2, Value :3 }]
+        internal static TableValue TableFromJsonArray(JsonElement array)
+        {
+            Contract.Assert(array.ValueKind == JsonValueKind.Array);
+
+            var records = new List<RecordValue>();
+
+            for (var i = 0; i < array.GetArrayLength(); ++i)
+            {
+                var element = array[i];
+                var val = GuaranteeRecord(FromJson(element));
```

Comment:
>GuaranteeRecord(FromJson(element)) [](http://example.com/codeflow?start=26&length=34)

This would work with things like `[1, { Value: 2 }, 3]`, which we probably want to block